### PR TITLE
[Process] Add callback for waiting process execution.

### PR DESCRIPTION
**Pull request by notdefine**
…

| Q             | A
| ------------- | ---
| Branch?       | 6.1 for features  
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       |  #... 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

-----

Add the possibility to add a callback while waiting for the process to finish. This allows to add Logging or other actions.

```php
<?php

class ExampleUsage
{
    private Logger $logger;

    public function runMyTask()
    {
        $process = new Symfony\Component\Process\Process('ffmpeg ...');
        $process->wait(
            null,
            function (Symfony\Component\Process\Process $process): void {
                $this->logger->notice(
                    'Ffmpeg says: ' . $process->getIncrementalErrorOutput()
                );
            }
        );
    }
}

```

### CODE REVIEW

**Comment by notdefine** [comment]
> It looks like you unchecked the "Allow edits from maintainer" box. That is fine, but please note that if you have multiple commits, you'll need to squash your commits into one before this can be merged. Or, you can check the "Allow edits from maintainers" box and the maintainer can squash for you.
> 
> Cheers!
> 
> Carsonbot

I did not found this setting :( Even with the reading of https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork - I dont see this option.

**Comment by fancyweb** [comment]
Hello, my question is naive but can't you use the existing callback and do the exact same thing? What's the difference here? Can you explain how the existing callback is limiting you? Thanks.

**Comment by nicolas-grekas** [comment]
I feel like this is better covered by using `foreach ($process as $type => $data)`, see https://symfony.com/doc/current/components/process.html

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 1]
```diff
@@ -1535,6 +1535,18 @@ public function testEnvCaseInsensitiveOnWindows()
         }
     }
 
+    public function testProcessWaitingCallback()
+    {
+        $p = $this->getProcessForCode('usleep(3000000);');
+        $callCounter = 0;
+        $p->wait(null,
+            function ($type, $buffer) use (&$callCounter) {
```
Comment (derrabus) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: Why `$type` and `$buffer`? Your implementation passes the actual `Process` instance to the callback.

File: src/Symfony/Component/Process/Process.php [hunk 2]
```diff
@@ -397,15 +397,16 @@ public function restart(callable $callback = null, array $env = []): static
...
+     * @param callable|null $waitingCallback A valid PHP callback triggered while waiting
      *
      * @return int The exitcode of the process
      *
      * @throws ProcessTimedOutException When process timed out
      * @throws ProcessSignaledException When process stopped after receiving signal
      * @throws LogicException           When process is not yet started
      */
-    public function wait(callable $callback = null): int
+    public function wait(callable $callback = null, callable $waitingCallback = null): int
```
Comment (derrabus) [inline src/Symfony/Component/Process/Process.php:409]: Adding a new parameter (even an optional one) to a non-final method is a breaking change. You need to comment out the parameter and use `func_num_args()`/`func_get_arg()` to retrieve it.

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 3]
```diff
@@ -1535,6 +1535,18 @@ public function testEnvCaseInsensitiveOnWindows()
...
+                ++$callCounter;
+            }
+        );
+        $this->assertSame(2, $callCounter, 'The callback should be executed while waiting');
```
Comment (derrabus) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: Can you elaborate why you expect your callback to be called exactly twice?

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by notdefine** [comment]
> Hello, my question is naive but can't you use the existing callback and do the exact same thing? What's the difference here? Can you explain how the existing callback is limiting you? Thanks.

The current Callback is only called when the Process is finished. The new one while the process is running.

File: src/Symfony/Component/Process/Tests/ProcessTest.php (hunk 3 above)
Comment (notdefine) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: The executed process is a sleep for 3 seconds. And the Callback is called every second.

**Comment by nicolas-grekas** [comment]
> I feel like this is better covered by using foreach ($process as $type => $data)

Closing as this stalled and this objection remains.

Thanks for proposing.

**Pull request closed without merging**